nx/lib/nx.ex

```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
+      iex> weights = Nx.tensor([[[8, 2, 4], [0, 3, 5]], [[4, 2, 6], [3, 1, 5]]])
+      iex> Nx.weighted_mean(t, weights, axis: :x)
+      #Nx.Tensor<
+        f32[y: 2][z: 3]
+        [
+          [3.0, 5.0, 6.599999904632568],
+          [10.0, 6.5, 9.0]
+        ]
+      >
+
+      iex> t = Nx.tensor([[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]], names: [:x, :y, :z])
+      iex> weights = Nx.tensor([[[8, 2, 4], [0, 3, 5]], [[4, 2, 6], [3, 1, 5]]])
+      iex> Nx.weighted_mean(t, weights, axis: -1)
+      #Nx.Tensor<
+        f32[x: 2][y: 2]
+        [
+          [1.7142857313156128, 5.625],
+          [8.166666984558105, 11.222222328186035]
+        ]
+      >
+
+  ### Keeping axis
+
+      iex> t = Nx.tensor([[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]], names: [:x, :y, :z])
+      iex> weights = Nx.tensor([[[8, 2, 4], [0, 3, 5]], [[4, 2, 6], [3, 1, 5]]])
+      iex> Nx.weighted_mean(t, weights, axis: -1, keep_axis: true)
+      #Nx.Tensor<
+        f32[x: 2][y: 2][z: 1]
+        [
+          [
+            [1.7142857313156128],
+            [5.625]
+          ],
+          [
+            [8.166666984558105],
+            [11.222222328186035]
+          ]
+        ]
+      >
+  """
+  @doc type: :aggregation, from_backend: false
+  def weighted_mean(tensor, weights, opts \\ []) do
+    opts = keyword!(opts, axis: nil, keep_axis: false)
+    %T{shape: shape, names: names} = tensor = to_tensor(tensor)
+    %T{shape: weights_shape} = weights = to_tensor(weights)
+
+    axis =
+      if opts[:axis] != nil,
+        do: Nx.Shape.normalize_axes(shape, [opts[:axis]], names),
+        else: opts[:axis]
+
+    weights =
+      if shape != weights_shape do
+        cond do
+          axis == nil ->
+            raise ArgumentError, "Axis must be specified when shapes of input and weights differ."
+
+          tuple_size(weights_shape) != 1 ->
+            raise ArgumentError, "1D weights expected when shapes of a and weights differ."
+
+          elem(weights_shape, 0) != elem(shape, axis) ->
+            raise ArgumentError, "Length of weights not compatible with specified axis."
+
+          true ->
+            nil
+        end
+
+        dims_to_broadcast =
+          for(_ <- 0..(tuple_size(shape) - 2), do: 1) ++ Tuple.to_list(weights_shape)
+
+        dims_to_broadcast = List.to_tuple(dims_to_broadcast)
+        weights = broadcast(weights, dims_to_broadcast)
+        dims_to_swap = for i <- 0..(tuple_size(shape) + tuple_size(weights_shape) - 2), do: i
+
+        dims_to_swap = swap(dims_to_swap, axis, tuple_size(shape) + tuple_size(weights_shape) - 2)
+
+        transpose(weights, axes: dims_to_swap)
+      else
+        weights
+      end
+
+    weights_sum = sum(weights, axes: axis, keep_axes: opts[:keep_axis])
+
+    if to_number(any(equal(weights_sum, 0.0))) == 1 do
+      raise ArgumentError, "Weights sum to zero, can't be normalized"
+    end
```

Comment:
I would just let it crash in the operation. This check will take the operation from the device instead of running everything in-device